server/mlabns/handlers/update.py

```diff
@@ -214,57 +214,59 @@ def get(self):
                                   site_id=site_id).get()
             if site == None:
                 logging.info('mlab-ns does not support site %s.', site_id)
                 continue
 
-            sliver_tool_gql = model.SliverTool.gql('WHERE fqdn=:fqdn',
-                                                   fqdn=fqdn)
-
-            if sliver_tool_gql.count() > 0:
-                # FQDN is not necessarily unique across tools.
-                for sliver_tool in sliver_tool_gql.run(
-                        batch_size=constants.GQL_BATCH_SIZE):
-                    # If the ipv4/6 addresses at self.IP_LIST_URL differ from
-                    # what is in the datastore then update the datastore
-                    if not (sliver_tool.sliver_ipv4 == ipv4 and
-                            sliver_tool.sliver_ipv6 == ipv6):
-                        sliver_tool = self.set_ip_sliver_tool(sliver_tool, ipv4,
-                                                              ipv6)
-
-                        self.put_sliver_tool(sliver_tool)
-            else:
-                # Sliver tool does not exist in datastore. Initialize sliver if
-                # the corresponding tool exists in the Tool table and the
-                # corresponding site exists in the Site table. This case occurs
-                # when a new tool has been added after the last IPUpdateHanlder
-                # ran. The sliver tool will actually be written to datastore at
-                # the next step.
-                logging.warning('sliver_tool %s is not in datastore.', fqdn)
-
-                sliver_tool = self.initialize_sliver_tool(tool, site, server_id,
-                                                          fqdn)
+            # If mlab-ns does not serve/support this slice, then skip it. Note:
+            # a given slice_id might have multiple tools (e.g., iupui_ndt has
+            # both 'ndt' and 'ndt_ssl' tools.
+            tools = model.Tool.gql('WHERE slice_id=:slice_id',
+                                   slice_id=slice_id)
+            if tools.count() == 0:
+                continue
 
-                sliver_tool = self.set_ip_sliver_tool(sliver_tool, ipv4, ipv6)
+            for tool in tools.run():
+                # Query the datastore to see if this sliver_tool exists there.
+                sliver_tool_gql = model.SliverTool.gql(
+                    'WHERE fqdn=:fqdn AND tool_id=:tool_id',
+                    fqdn=fqdn,
+                    tool_id=tool.tool_id)
+
+                # Check to see if the sliver_tool already exists in the
+                # datastore. If not, add it to the datastore.
+                if sliver_tool_gql.count() == 1:
+                    sliver_tool = sliver_tool_gql.get(
+                        batch_size=constants.GQL_BATCH_SIZE)
+                elif sliver_tool_gql.count() == 0:
+                    logging.info(
+                        'For tool {}, fqdn {} is not in datastore.  Adding it.'.format(
```

Comment:
prefer 
```python
logging.info('For tool %s, fqdn %s is not in datastore.  Adding it.', tool.tool_id, fqdn)
```
because then if the log level is such that INFO messages are not printed, then the string formatting work will not be done. This is a tiny optimization, and doesn't matter here, so LGTM either way.